cfmm.mligo

```diff
@@ -293,6 +303,87 @@ let cash_transfer (storage : storage) (from : address) (to_ : address) (cash_amo
 #endif
 #endif
 
+(* Isoutility and Difference Equations *)
+// The Isoutility Function in https://hackmd.io/MkPSYXDsTf-giBprcDrc3w  
+let isoutility (target, cash, token : (nat * nat) * nat * nat) : nat = 
+    let x = cash in 
+    let y = token in 
+    let (a,b) = target in 
+    let a2 = a * a in
+    let b2 = b * b in 
+    let ax2 = a2 * x * x in 
+    let by2 = b2 * y * y in
+    abs (a * x * b * y * (ax2 + by2) / (2 * a2 * b2))
+
+// Returns the price dy/dx of the isoutility function, i.e. a map by multiplication ∆x => ∆y, at a given point (x,y)
+let price_cash_to_token (target : nat * nat) (cash : nat) (token : nat) : nat = 
+    let (x,y) = (cash, token) in
+    let (a,b) = target in 
+    let ax2 = x * x * a * a in
+    let by2 = y * y * b * b in
+    let num = y * (3n * ax2 + by2) in
+    let denom = x * (ax2 + 3n * by2) in
+    num/denom
+
+// A function to transfer assets while maintaining a constant isoutility
+let rec newton_dx_to_dy (x, y, dx, dy_approx, target, rounds : nat * nat * nat * nat * (nat * nat) * (nat option)) : nat = 
+    let (a,b) = target in 
+    let xp = x + dx in
+    let yp = y - dy_approx in 
+    let ax2 = a * a * x * x in let by2 = b * b * y * y in 
+    let axp2 = a * a * xp * xp in let byp2 = b * b * yp * yp in
+    (* Newton descent formulae *)
+    let num = x * y * (ax2 + by2) - xp * yp * (axp2 + byp2) in 
+    let denom = xp * (axp2 + 3 * byp2) in
+    let adjust = num / denom in 
+    // if (abs adjust <= 1n) (* marginal difference calculated is <= 1mutez *)
+    let n = 
+        (match rounds with
+        | None -> 4n
+        | Some n -> n)
+    in
+    if (n <= 0n) (* Newton converges in 4 rounds, so we bound computation there *)
+    then 
+        let dy = dy_approx - adjust in 
+        if y - dy <= 0 
+        then
+            (failwith error_TOKEN_POOL_MINUS_TOKENS_WITHDRAWN_IS_NEGATIVE : nat)
```

Comment:
this check shouldn't be necessary, mathematically won't happen since dy remains an underestimate

Reply:
Yeah you're right, I think that coupled with the checks done in `trade_dcash_for_dtoken` and `trade_dtoken_for_dcash` make this redundant. Removed.